backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py

```diff
@@ -0,0 +1,153 @@
+         or should it be scheduled for the future.
+
+        :return: pass dictionary with job_delay in seconds to the next handler
+        """
+        record = list(mongo_config_collection.find())[0]
+        last_update = record["previous_job_start_time"]
+        if last_update is None:
+            # If it's the first time that the job is run (last_update is None) then
+            # job delay should be equal to CHANGES_INTERVAL_SECONDS. Update the mongo record with job state accordingly.
+            job_delay = CHANGES_INTERVAL_SECONDS
+            mongo_config_collection.update_one({"_id": record["_id"]},
+                                               {"$set": {"previous_job_start_time": datetime.datetime.utcnow()}})
+        else:
+            # Check how many seconds have elapsed since the last time that the job was run. If the time difference
+            # is greater than CHANGES_INTERVAL_SECONDS then job can be run immediately. Otherwise, calculate how
+            # many seconds are left until minimum time difference between updates (CHANGES_INTERVAL_SECONDS).
+            current_time = datetime.datetime.utcnow()
+            delta = current_time - last_update
+            time_diff = delta.total_seconds()
+            if time_diff > CHANGES_INTERVAL_SECONDS:
+                job_delay = 1
+            else:
+                job_delay = int(CHANGES_INTERVAL_SECONDS-time_diff)
+
+        result = {
+            "job_delay": job_delay
+        }
+
+        return super().handle(result)
+
+class ScheduleHandler(AbstractHandler):
+    def handle(self, request: dict):
+        """
+        ScheduleHandler schedules the kubernetes job with updated sc4snmp configuration
+        """
+        record = list(mongo_config_collection.find())[0]
+        if not record["currently_scheduled"]:
+            # If records isn't currently scheduled, schedule it and update its state in mongo
+            mongo_config_collection.update_one({"_id": record["_id"]},
+                                         {"$set": {"currently_scheduled": True}})
+            run_job.apply_async(countdown=request["job_delay"], queue='apply_changes')
+        return request["job_delay"]
+
+
+class SingletonMeta(type):
+    _instances = {}
+    _lock: Lock = Lock()
+
+    def __call__(cls, *args, **kwargs):
+        with cls._lock:
+            if cls not in cls._instances:
+                instance = super().__call__(*args, **kwargs)
+                cls._instances[cls] = instance
+        return cls._instances[cls]
+
+class ApplyChanges(metaclass=SingletonMeta):
+    def __init__(self) -> None:
+        """
+        ApplyChanges is a singleton responsible for creating mongo record with a current state of kubernetes job.
+        """
+        self.__handling_chain = CheckJobHandler()
+        schedule_handler = ScheduleHandler()
+        self.__handling_chain.set_next(schedule_handler)
+        if len(list(mongo_config_collection.find({}))) == 0:
```

Comment:
Is there any reason why it can't be `if not mongo_config_collection.find({})`

Reply:
<img width="384" alt="Screenshot 2023-06-24 at 20 42 44" src="https://github.com/splunk/SC4SNMP-UI/assets/108660584/a9d285ea-3b26-4150-828f-e32ac0634132">
<img width="385" alt="Screenshot 2023-06-24 at 20 42 49" src="https://github.com/splunk/SC4SNMP-UI/assets/108660584/d4ac3b16-3927-4452-8cda-562a7f7b15c3">

Reply:
As you can see `if not mongo_config_collection.find({})` will return `False` even if  collection is empty

Reply:
ok, but `if not list(mongo_config_collection.find({})):` will work, right?

Reply:
Ok, I changed the code a little bit and now `upsert` is used. Now there is no `if len(list(mongo_config_collection.find({}))) == 0:` line